In [12]:
import pandas as pd

# read csv file into a DataFrame, but not 1st lines
df = pd.read_csv('/mnt/c/Users/minhhoang/ztubt_placedbettransactionheader.csv')
# assign null to a new column
df['X_RECORD_INSERT_TS'] = None
df['X_RECORD_UPDATE_TS'] = None
df['X_ETL_NAME'] = None
df

tranheaderid                          requestid  \
0     0322458D-6FC9-4CCA-9580-28AE0208289A                1511890010001355370   
1     0B637CB2-C878-4909-8DD5-2F5A434F459A                1511890010001355439   
2     0FD876B4-4A2A-4AEA-B772-56D07290B80E                1511890010001355317   
3     13BA0BA1-06E4-429A-A3F1-F9D882656B48                1511890010001355314   
4     14D2C076-47E6-4A6A-8306-503E88C1E78E                1511890010001355339   
...                                    ...                                ...   
4799  B9E30833-3379-4B56-B736-7836182B0135  0088880340000135824,1764812852795   
4800  EB2B62C3-C5B0-4A90-BBFE-CCF1AA45C175  0088880340000135840,1764813595463   
4801  7365BCA5-2381-4FF0-BE6A-C759473EA87D  0088880340000135841,1764813713937   
4802  AD55F4A5-B420-4E56-B6FC-528636519110  0088880340000135842,1764819500598   
4803  E3E05759-BC03-436B-B3D2-5AE4CACEB613                0088880340000135843   

                               tickettransactionid  \
0     mXF9PTKe7GDbrH/9wHSkRqYUHvSaKieZlpkSwguNiTw=   
1     UUZ7nivjtEdI4aqWfxpEw9xj+Ox7W7C21UaNf8K5MnQ=   
2     qTH44nXVO3LkJcvuoOGl37yNd9xczwC3RW22tJFD9Pg=   
3     +DwnDsDzm+GtXbMeBSGpxkFggNedfZooRjOiqd17OsQ=   
4     +3k0DL9Oq4uDblegw4ltXC7bcjZja1n8yUQWZESGfq0=   
...                                            ...   
4799  QOpQoZjCgAVKV3YGHphYdTiBIhy74IIOqa5usaSZSLw=   
4800  W1jdh+47sHuT6BF8FuhwcvygGhD+OGsRQ1konxYnIzE=   
4801  lTPXmiKSenUbvk2mm9LGKXTNfLZQsU5aOPUSF2pEvMA=   
4802  ybtizQY2T0TnK5q9w3pOvKHf6HE0ZtsFZRnLosYCcgg=   
4803  zJgGH6zy5jHtlw8MVlLpPY8hVqQGPbV05IHrQashbu4=   

                                ticketserialnumber         hostticketid  \
0     Xcfo6qfXyoRyHS+0PuHOiM40nAfSdz2tPqp9mE0Cr8k=             41244327   
1     kjmeBXZmjle7UKlKP2fTnBT3fQcjvxpam5PBrnjQpvs=             41244395   
2     hRRgLLVaZawpe4V6oyxVpYxy48ieC1o9lvlSUdLc3N4=             41243965   
3     PGI1zaDOCx8Md8rcxvkbA5jVKEYP8MHus1yQ3G/PYT8=             41243908   
4     FMi4KOqRWCJX0PeSb3xC/A2vocktsLp2KU+MJUEwSGc=             41244147   
...                                            ...                  ...   
4799  lE2+yIS21dRwmH2JxBYRQfqeTSpuh3Hhr6KVOBAx+tI=  088-26719232-233008   
4800  eipydWOsOOAmar9uB7jGbm+5dX6XX5IvVCcp1ASK8os=  088-47628544-235408   
4801  SZaThP3W3rXxwXWpZTGwaDfmgDM9sd85kZ1UxuS/B2c=  088-16030977-090408   
4802  ivwrlm4zloL9Pu3h+IbEMOX3HdUuLb+GzHawHW1Podg=  088-60549376-113108   
4803  AleFaFhMp+vPuhI59VMids4trZj5h73MpjZD0GScuGs=             42164137   

      barcode             trandate  prodid  \
0         NaN  2025-06-18 15:01:10       1   
1         NaN  2025-06-18 15:43:56       1   
2         NaN  2025-06-18 14:20:35       1   
3         NaN  2025-06-18 14:18:14       1   
4         NaN  2025-06-18 14:32:25       1   
...       ...                  ...     ...   
4799      NaN  2024-12-05 09:47:33       3   
4800      NaN  2024-12-05 09:59:55       3   
4801      NaN  2024-12-05 10:01:54       2   
4802      NaN  2024-12-05 11:38:20       4   
4803      NaN  2025-12-04 11:38:41       1   

                                    cartid          createddate  ...  ispay  \
0     63CB4FC6-0982-45D6-8323-F777D93A352A  2025-06-18 07:01:10  ...  False   
1     6D25608C-4C84-431B-B486-0A964C9FB783  2025-06-18 07:43:55  ...  False   
2     63CB4FC6-0982-45D6-8323-F777D93A352A  2025-06-18 06:20:35  ...  False   
3     63CB4FC6-0982-45D6-8323-F777D93A352A  2025-06-18 06:18:13  ...   True   
4     63CB4FC6-0982-45D6-8323-F777D93A352A  2025-06-18 06:32:24  ...  False   
...                                    ...                  ...  ...    ...   
4799  82E1B3C0-9CAE-457E-85E4-CFDCD40BDFF4  2025-12-04 01:47:32  ...  False   
4800  82E1B3C0-9CAE-457E-85E4-CFDCD40BDFF4  2025-12-04 01:59:55  ...  False   
4801  82E1B3C0-9CAE-457E-85E4-CFDCD40BDFF4  2025-12-04 02:01:54  ...  False   
4802  82E1B3C0-9CAE-457E-85E4-CFDCD40BDFF4  2025-12-04 03:38:20  ...  False   
4803  82E1B3C0-9CAE-457E-85E4-CFDCD40BDFF4  2025-12-04 03:38:41  ...  False   

     

In [14]:
# Kiểm tra tên cột và thông tin DataFrame
print("Tên các cột:")
print(df.columns.tolist())
print("\nThông tin DataFrame:")
print(df.info())
print("\nKích thước DataFrame:")
print(f"Rows: {len(df)}, Columns: {len(df.columns)}")
print("\n5 dòng đầu:")
print(df.head())

Tên các cột:
['tranheaderid', 'requestid', 'tickettransactionid', 'ticketserialnumber', 'hostticketid', 'barcode', 'trandate', 'prodid', 'cartid', 'createddate', 'entrymethodid', 'qrformat1', 'qrformat2', 'deviceid', 'isbetrejectedbytrader', 'ispay', 'uniqueid', 'iscancelled', 'isexchangeticket', 'terdisplayid', 'userdisplayid', 'sessionid', 'X_RECORD_INSERT_TS', 'X_RECORD_UPDATE_TS', 'X_ETL_NAME']

Thông tin DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4804 entries, 0 to 4803
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   tranheaderid           4804 non-null   object 
 1   requestid              4804 non-null   object 
 2   tickettransactionid    4804 non-null   object 
 3   ticketserialnumber     4804 non-null   object 
 4   hostticketid           4687 non-null   object 
 5   barcode                0 non-null      float64
 6   trandate               4708 non-null   object 
 

In [15]:
# Làm sạch tên cột để tránh lỗi với Snowflake
def clean_column_names(df):
    """Làm sạch tên cột để tương thích với Snowflake"""
    # Tạo một bản sao để không thay đổi DataFrame gốc
    cleaned_df = df.copy()

    # Làm sạch tên cột
    new_columns = []
    for col in cleaned_df.columns:
        # Loại bỏ ký tự đặc biệt, chuyển thành chữ hoa
        clean_col = str(col).strip().upper()
        # Thay thế khoảng trắng và ký tự đặc biệt bằng underscore
        import re
        clean_col = re.sub(r'[^A-Z0-9_]', '_', clean_col)
        # Loại bỏ underscore liên tiếp
        clean_col = re.sub(r'_+', '_', clean_col)
        # Loại bỏ underscore ở đầu và cuối
        clean_col = clean_col.strip('_')
        new_columns.append(clean_col)

    cleaned_df.columns = new_columns
    return cleaned_df

# Áp dụng làm sạch tên cột
df_clean = clean_column_names(df)
print("Tên cột sau khi làm sạch:")
print(df_clean.columns.tolist())
print("\nKiểm tra vài dòng đầu của dữ liệu đã làm sạch:")
df_clean.head()

Tên cột sau khi làm sạch:
['TRANHEADERID', 'REQUESTID', 'TICKETTRANSACTIONID', 'TICKETSERIALNUMBER', 'HOSTTICKETID', 'BARCODE', 'TRANDATE', 'PRODID', 'CARTID', 'CREATEDDATE', 'ENTRYMETHODID', 'QRFORMAT1', 'QRFORMAT2', 'DEVICEID', 'ISBETREJECTEDBYTRADER', 'ISPAY', 'UNIQUEID', 'ISCANCELLED', 'ISEXCHANGETICKET', 'TERDISPLAYID', 'USERDISPLAYID', 'SESSIONID', 'X_RECORD_INSERT_TS', 'X_RECORD_UPDATE_TS', 'X_ETL_NAME']

Kiểm tra vài dòng đầu của dữ liệu đã làm sạch:


TRANHEADERID            REQUESTID  \
0  0322458D-6FC9-4CCA-9580-28AE0208289A  1511890010001355370   
1  0B637CB2-C878-4909-8DD5-2F5A434F459A  1511890010001355439   
2  0FD876B4-4A2A-4AEA-B772-56D07290B80E  1511890010001355317   
3  13BA0BA1-06E4-429A-A3F1-F9D882656B48  1511890010001355314   
4  14D2C076-47E6-4A6A-8306-503E88C1E78E  1511890010001355339   

                            TICKETTRANSACTIONID  \
0  mXF9PTKe7GDbrH/9wHSkRqYUHvSaKieZlpkSwguNiTw=   
1  UUZ7nivjtEdI4aqWfxpEw9xj+Ox7W7C21UaNf8K5MnQ=   
2  qTH44nXVO3LkJcvuoOGl37yNd9xczwC3RW22tJFD9Pg=   
3  +DwnDsDzm+GtXbMeBSGpxkFggNedfZooRjOiqd17OsQ=   
4  +3k0DL9Oq4uDblegw4ltXC7bcjZja1n8yUQWZESGfq0=   

                             TICKETSERIALNUMBER HOSTTICKETID  BARCODE  \
0  Xcfo6qfXyoRyHS+0PuHOiM40nAfSdz2tPqp9mE0Cr8k=     41244327      NaN   
1  kjmeBXZmjle7UKlKP2fTnBT3fQcjvxpam5PBrnjQpvs=     41244395      NaN   
2  hRRgLLVaZawpe4V6oyxVpYxy48ieC1o9lvlSUdLc3N4=     41243965      NaN   
3  PGI1zaDOCx8Md8rcxvkbA5jVKEYP8MHus1yQ3G/PYT8=     41243908      NaN   
4  FMi4KOqRWCJX0PeSb3xC/A2vocktsLp2KU+MJUEwSGc=     41244147      NaN   

              TRANDATE  PRODID                                CARTID  \
0  2025-06-18 15:01:10       1  63CB4FC6-0982-45D6-8323-F777D93A352A   
1  2025-06-18 15:43:56       1  6D25608C-4C84-431B-B486-0A964C9FB783   
2  2025-06-18 14:20:35       1  63CB4FC6-0982-45D6-8323-F777D93A352A   
3  2025-06-18 14:18:14       1  63CB4FC6-0982-45D6-8323-F777D93A352A   
4  2025-06-18 14:32:25       1  63CB4FC6-0982-45D6-8323-F777D93A352A   

           CREATEDDATE  ...  ISPAY             UNIQUEID ISCANCELLED  \
0  2025-06-18 07:01:10  ...  False  1511890010001355370       False   
1  2025-06-18 07:43:55  ...  False  1511890010001355439       False   
2  2025-06-18 06:20:35  ...  False  1511890010001355317       False   
3  2025-06-18 06:18:13  ...   True  1511890010001355314       False   
4  2025-06-18 06:32:24  ...  False  1511890010001355339       False   

  ISEXCHANGETICKET  TERDISPLAYID USERDISPLAYID  \
0            False     151189001      S0004948   
1            False     151189001      S0004948   
2            False     151189001      S0004948   
3            False     151189001      S0004948   
4            False     151189001      S0004948   

                              SESSIONID  X_RECORD_INSERT_TS  \
0  613ACF4F-732B-4F45-B2A1-BE597A9DC65C                None   
1  613ACF4F-732B-4F45-B2A1-BE597A9DC65C                None   
2  613ACF4F-732B-4F45-B2A1-BE597A9DC65C                None   
3  613ACF4F-732B-4F45-B2A1-BE597A9DC65C                None   
4  613ACF4F-732B-4F45-B2A1-BE597A9DC65C                None   

   X_RECORD_UPDATE_TS  X_ETL_NAME  
0                None        None  
1                None        None  
2                None        None  
3                None        None  
4                None        None  

[5 rows x 25 columns]

In [16]:
from write_pandas import write_to_snowflake

from Snowflake_connection import snowflake_connection

conn = snowflake_connection()
database = 'SPPL_DEV_DWH'
schema = 'SPPL_PUBLIC'
table_name = 'ZTUBT_PLACEDBETTRANSACTIONHEADER'

# Sử dụng DataFrame đã làm sạch thay vì DataFrame gốc
write_to_snowflake(df_clean, conn, database=database, schema=schema, table_name=table_name)

Current context - Database: SPPL_DEV_DWH, Schema: SPPL_PUBLIC, Warehouse: SPPL_WH_DEV
Switched to Database: SPPL_DEV_DWH, Schema: SPPL_PUBLIC
Table SPPL_DEV_DWH.SPPL_PUBLIC.ZTUBT_PLACEDBETTRANSACTIONHEADER exists. Clearing existing data...
Inserting new data into existing table...
Verification: 4804 rows found in SPPL_DEV_DWH.SPPL_PUBLIC.ZTUBT_PLACEDBETTRANSACTIONHEADER
Successfully written 4804 rows to SPPL_DEV_DWH.SPPL_PUBLIC.ZTUBT_PLACEDBETTRANSACTIONHEADER
Number of chunks: 1, Success: True


## Tóm tắt quá trình xử lý dữ liệu

✅ **Đã hoàn thành thành công!**

1. **Load dữ liệu**: Đã đọc thành công file CSV với 4804 dòng dữ liệu
2. **Xử lý lỗi tên cột**: Đã làm sạch tên cột để tương thích với Snowflake
   - Chuyển tất cả tên cột sang chữ hoa
   - Loại bỏ ký tự đặc biệt và thay thế bằng underscore
3. **Thêm metadata**: Đã thêm các cột hệ thống:
   - X_RECORD_INSERT_TS
   - X_RECORD_UPDATE_TS  
   - X_ETL_NAME
4. **Ghi dữ liệu**: Đã ghi thành công 4804 dòng vào bảng `SPPL_DEV_DWH.SPPL_PUBLIC.ZTUBT_PLACEDBETTRANSACTIONHEADER`

**Lỗi đã sửa**: Invalid identifier "tranheaderid" → Đã chuẩn hóa tên cột thành "TRANHEADERID"